### DISCUSSION TITLE
For large page props random JSON parse error on client side

### DISCUSSION THREAD

**Discussion by jorups** [vercel/next.js#41299] [Help] [answered]
We have huge page props (~3.5MB, it's not ok, will find ways to decrease it) and randomly (not always) we are getting JSON.parse exception.
```
SyntaxError: Unterminated string in JSON at position 2849123
    at JSON.parse (<anonymous>)
    at Object.<anonymous> (main-1257a2700f0e35f3.js:1:15365)
```
This exception is raised on the client side at initialize() https://github.com/vercel/next.js/blob/e933e1d211ea16c349898e141d9733b4cd06e3d8/packages/next/client/index.tsx#L199

Interestingly position where JSON.parse fails varies. When trying to access `__NEXT_DATA__` dom element from js console `(console.dir(document.getElementById("__NEXT_DATA__").textContent)` JSON is correct and parsable. 

This makes me think that client's initialize() is called when DOM is not fully loaded. Script tag with id `__NEXT_DATA__` is at the bottom of the retrieved HTML so might be that as it's very large, it's not yet fully loaded. 

Is that expected behavior?

**Comment by icyJoseph** [comment, marked as answer]
Interesting find, I'd be curious to know more. I will try to figure out who calls `initialize`, but I can tell you already that when I added this to my `_app`:

```ts
if (typeof window !== "undefined") {
  document.addEventListener("DOMContentLoaded", () => {
    console.log("DOM fully loaded and parsed");
  });
  window.addEventListener("load", () => console.log("page is fully loaded"));
}
```

And then modified the source of `intialize` to print to console when it is called, then I got:

```shell
initialize
DOM fully loaded and parsed
page is fully loaded
```

`console` logging is not the best, but I guess here it shows that you might be onto something! Perhaps worth adding that I've observed this on dev and production builds.

## Edit

Funny, I had been tinkering with the 'run-functions' on start-up discussion and did work a bit with `next-dev` and `next`:

- packages/next/client/next.js
- packages/next/client/next-dev.js

And this makes sense because these scripts load with [`defer` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-defer), which:

> This Boolean attribute is set to indicate to a browser that the script is meant to be executed after the document has been parsed, but before firing [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event).

So the question is, if the document has been parsed, why is the JSON.parse call failing at that moment. The plot thickens. Are you sure your JSON is always correctly generated? Are there other parties writing to it?

## Resolution

This is not a problem with Next.js, but with how browsers are transitioning between readyState loading, to interactive, and deferred scripts runtime, when there's a large DOM element (3.5MB in this case) on the page. 

Through our discussion OP was able to produce, https://gist.github.com/jorups/cd7101dbd0006a282e0abe31823aa0c9, which showcases the issue very clearly by having a simple HTML file and an external script.

**Reply by jorups** [reply]
Pages are pre-generated (SSG), so I can see the full HTML with `__NEXT_DATA__` server side, they're complete and parsable. And I know that content of pre-generated pages is not modified during refreshes when I'm getting JSON.parse errors. 

Where do you observe that these scripts are loaded with `defer` attribute?

**Reply by icyJoseph** [reply]
<img width="584" alt="Screenshot 2022-10-10 at 10 53 49" src="https://user-images.githubusercontent.com/21013447/194830183-05d7f088-e19b-4a19-8a2f-44e277fba0a4.png">

I inspected the DOM, on various projects, in dev and production mode, and it checks out, they have `defer` attribute set.

Well, I wonder if any browser extension could be getting in the way? Really odd problem. I guess you could add your own `load`, or `DOMContentLoaded`, event listener as well, and try to parse __NEXT_DATA__ from within `_app`.

**Reply by jorups** [reply]
Thanks for your reply!
I added `console.log(document.readyState)` just before 
```
initialData = JSON.parse(
    document.getElementById('__NEXT_DATA__')!.textContent!
  )
```
Interestingly, JSON.parse failed when document.readyState was `loading` but succeeded (no JSON parse errors) when state was `interactive`.

So in docs (https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState) it states that:
```
case "loading":
    // The document is still loading.
    break;
case "interactive": {
    // The document has finished loading. We can now access the DOM elements.
    // But sub-resources such as scripts, images, stylesheets, and frames are still loading.
case "complete":
    // The page is fully loaded.
```

Seems these scripts are run when the document is still loading and not fully ready yet.

Don't know but might it be the case that 'document has been parsed' in the statement below means that it was parsed (syntactically) but not yet loaded into the DOM and therefore not available for DOM queries?
`This Boolean attribute is set to indicate to a browser that the script is meant to be executed after the document has been parsed, but before firing [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event).`

**Reply by icyJoseph** [reply]
Are you observing this across browsers? Or only in a particular one? If so which version?

**Reply by jorups** [reply]
Across browsers - Chrome, Safari.

I think I found the culprit for the problem, but can't explain the behavior yet :)

I'm using `@next-safe/middleware` package which enables CSP for Next.js framework. For SSG-type pages it uses hash-based CSP and what it does is that it rewrites static <script> definitions to dynamic, hence following `strict-dynamic` CSP:
```
<script src="/_next/static/chunks/webpack-bbd44435dc4b036d.js" defer=""></script>
<script src="/_next/static/chunks/framework-5f4595e5518b5600.js" defer=""></script>
<script src="/_next/static/chunks/main-a18e6306682495f5.js" defer=""></script>
```

becomes

```
<script id="rO0mn1vmdEmchPfbkztT+xiCgJtEq4hz1ca8n34RE7I=" integrity="sha256-UudEbvkUR0NywEde62Tu/Pp/KaKVnxWk2XUQzM9hAhU=">
            (function() {
                var s0 = document.createElement('script');
                s0.async = false;
                s0.defer = true;
                s0.src = '/_next/static/chunks/webpack-bbd44435dc4b036d.js';
                var s1 = document.createElement('script');
                s1.async = false;
                s1.defer = true;
                s1.src = '/_next/static/chunks/framework-5f4595e5518b5600.js';
                var s2 = document.createElement('script');
                s2.async = false;
                s2.defer = true;
                s2.src = '/_next/static/chunks/main-1257a2700f0e35f3.js';
                ........
                var s = [s0, s1, s2, ........];
                var self = document.getElementById('rO0mn1vmdEmchPfbkztT+xiCgJtEq4hz1ca8n34RE7I=');
                var p = self.parentNode;
                p.removeChild(self);
                s.forEach(function(si) {
                    p.appendChild(si);
                });
            }
            )()
        </script>
```

You can notice that script elements are created dynamically with `defer` attribute. However, somehow these scripts are executed before the DOM is fully loaded. For inline scripts, you can't add `defer` so they could be executed while DOM is not fully loaded (depends on the place in HTML), however, I'm not sure that the executor (browser) applies `defer` logic for dynamically added scripts with `defer` attribute.

**Reply by icyJoseph** [reply]
Weird indeed, but defer scripts are ran right before firing DOMContentLoaded:

> Scripts with the defer attribute will prevent the DOMContentLoaded event from firing until the script has loaded and finished evaluating.

That being said the `interactive` state is where we want to be. I ran a few tests, and could never get loading on my deferred script.

```js
(function(){
           const script = document.createElement("script");
           script.async = false;
           script.defer = true;
           script.src = "/onload.js" // this logs readyState interactive
           let self = document.getElementById("foo");
           let p = self.parentNode;
           p.removeChild(self);
           p.appendChild(script);
          })()
```

I wonder though if your plugin is messing with the application/json type script? putting an actual script in place? Because then the deferred scripts load in turn, and the generator of __NEXT_DATA__ script would be evaluated last. However, then this should fail all of the time. And still doesn't explain why the readyState is still loading.

**Reply by jorups** [reply]
Here is the gist with two files where I tried to reproduce this behavior, very similar to your snippet. (https://gist.github.com/jorups/cd7101dbd0006a282e0abe31823aa0c9)
At least for me (Chrome Version 105.0.5195.125 (Official Build) (arm64)) with dynamically added scripts I get an error and it writes 'loading' for readystate. When I try static script with `defer`, the script is executed and readystate is 'īnteractive' with no exceptions thrown. Disabled all extensions on Chrome so they don't interfere.

**Reply by icyJoseph** [reply]
Oh wow, that gist does trigger some funky stuff. Tested in Safari, FF, and Chrome. 

It might be that we are missing some kind of browser gotcha when it comes to loading -> interactive. The size of the `__NEXT_DATA__` JSON is clearly at fault here, but is this a bug? expected browser behavior?

This sounds a lot like a file handler that doesn't wait for the file to be fully written and immediately tries to close it. In a sense, as if it didn't await completion, and was just going on synchronously. 

With sufficiently small payloads it goes swell, but 3.5MB of JSON really make it cry.

At the end of the day it wasn't a Next.js problem though, for what is worth!

**Reply by jorups** [reply]
Yeah, weird stuff going on with huge DOM elements.
Thanks for your support and pleased that it's not Next.js fault!

**Reply by Meemaw** [reply]
Facing this as well trying to use `@next-safe/middleware`. I solved it by patching the package code by modifying the `isScriptElement` function (https://github.com/nibtime/next-safe-middleware/blob/c4075709fd8598ff3af27bafd25e0f298509dac5/packages/next-safe-middleware/src/document/csp-trustify/utils.tsx#L12), to return false if the script element is json:

```js
const isScriptElement = el => isJsxElement(el) && el.type === "script" && el.props.type !== "application/ld+json" && el.props.type !== "application/json";
```

JSON script elements are irrelevant for CSP as they don't execute javascript and should be kept alone.

However, after doing this I'm seeing this error:
<img width="686" alt="Screenshot 2023-06-02 at 21 11 48" src="https://github.com/vercel/next.js/assets/8524109/3db67c9a-f852-42e1-992a-ecd70dddf9be">

<img width="606" alt="Screenshot 2023-06-02 at 21 12 28" src="https://github.com/vercel/next.js/assets/8524109/1ef8653b-efed-4aaf-a3ec-66948df8627f">

It seems initialize is called, before the document is fully set up.

Fixed this with the following patch:

<img width="601" alt="Screenshot 2023-06-03 at 06 37 54" src="https://github.com/vercel/next.js/assets/8524109/a8a2f043-2a4a-46ad-8586-719496648be4">

All of this fells super hacky, so would be great if someone had any idea what is going on and if this is a code issue, or simply indeterministic browser behaviour.

**Reply by bimusiek** [reply]
We are having the same issue with large JSON data. (loading blog contents) Due to this we have removed next-safe package.
